geogen: introduce package to generate random geo types

otan opened the pull request:
Generate random geo types in a new geogen package. This is immediately
utilised for all the randomized testing, including sqlsmith.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51183)
<!-- Reviewable:end -->

rohany commented on pkg/geo/geogen/geogen_test.go:
```diff
@@ -0,0 +1,68 @@
+const numRuns = 25
+
+func TestRandomValidLinearRingCoords(t *testing.T) {
+	rng := rand.New(rand.NewSource(0))
```
following up from standup, I think you want to be using `randutil.NewPsuedoRand` so that your seed is different each time. Otherwise, this function does the same thing on every invocation.

otan replied:
I don't mind too much this is doing the same thing (it's a test that makes sure it generates something sane)
as long as the real random does the real thing.
that said, sure let's do it

sumeerbhola reviewed (approved):
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51183)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rohany)

___
*[pkg/geo/geogen/geogen.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGMV0IDcwSDIKd8rgC:-MDGMV0IDcwSDIKd8rgD:bnp15y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L43)):*
> ```Go
> // RandomValidLinearRingCoords generates a flat float64 array of coordinates that represents
> // a completely closed shape that can represent a simple LinearRing. This shape is always valid.
> // Implements the algorithm in https://observablehq.com/@tarte0/generate-random-simple-polygon.
> ```

can you add in the comment that numPoints should be >= 2
___
*[pkg/geo/geogen/geogen.go, line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGLwAaDem9ZWu0djLD:-MDGLwAaDem9ZWu0djLE:byx8af5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L63)):*
> ```Go
> 
> 	// Sort by the angle of all the points relative to the center.
> 	// Use descending order of angle to get a CCW loop.
> ```

doesn't descending angle make it CW (if my reading of the first 2 graphs in https://en.wikipedia.org/wiki/Atan2 is correct)? 
___
*[pkg/geo/geogen/geogen.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGNBXF3K2gd6Dvrwa5:-MDGNBXF3K2gd6Dvrwa6:bkbdck3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L92)):*
> ```Go
> 	randCoords := RandomValidLinearRingCoords(rng, numCoords, minX, maxX, minY, maxY)
> 
> 	// Truncate a random substring from the LineString.
> ```

How about s/Truncate/Extract/
___
*[pkg/geo/geogen/geogen.go, line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGNmhB7R56WOBXXXc3:-MDGNmhB7R56WOBXXXc4:b-v8jhle) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L114)):*
> ```Go
> 	//   and re-use "PointInLinearRing" to generate N random points inside the 4 points to form
> 	//   a "sub" linear ring inside.
> 	// * Generate a random set of polygons, see which ones they with fully cover and use that.
> ```

... they fully cover ...


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51183)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany and @sumeerbhola)

___
*[pkg/geo/geogen/geogen.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGMV0IDcwSDIKd8rgC:-MDGlF_w6CIS4XrcT07-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L43)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add in the comment that numPoints should be >= 2
</blockquote></details>

Done.
___
*[pkg/geo/geogen/geogen.go, line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGLwAaDem9ZWu0djLD:-MDGlUwq6A3slwkrISur:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L63)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

doesn't descending angle make it CW (if my reading of the first 2 graphs in https://en.wikipedia.org/wiki/Atan2 is correct)? 
</blockquote></details>

Done.
___
*[pkg/geo/geogen/geogen.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGNBXF3K2gd6Dvrwa5:-MDGlKyL1e2FBdSIx9ou:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L92)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

How about s/Truncate/Extract/
</blockquote></details>

Done.
___
*[pkg/geo/geogen/geogen.go, line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGNmhB7R56WOBXXXc3:-MDGlNaoE2FT-WmWoloN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L114)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

... they fully cover ...
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-:-MDHjr5pCSxGwEe6-l1d:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan, @rohany, and @sumeerbhola)

___
*[pkg/geo/geogen/geogen.go, line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGLwAaDem9ZWu0djLD:-MDHjWI1EWLu1kFPLEVY:bulsdao) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L63)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Was this not caught by a test because we later fix everything to be in CCW order?
___
*[pkg/geo/geogen/geogen.go, line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDHjEC59Fdvj3uorCet:-MDHjEC59Fdvj3uorCeu:b7f4bbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/954731d63dccb3d2119aa5a5caae6559c57f87a8/pkg/geo/geogen/geogen.go#L43)):*
> ```Go
> // RandomValidLinearRingCoords generates a flat float64 array of coordinates that represents
> // a completely closed shape that can represent a simple LinearRing. This shape is always valid.
> // A LinearRing must have at least 3 points. A point is added at the end to close the ring
> ```

nit: missing period


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51183)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan, @rohany, and @sumeerbhola)

___
*[pkg/geo/geogen/geogen.go, line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-MDGLwAaDem9ZWu0djLD:-MDHuUlo2rb6gNkgHaIA:b-hvb169) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cc3578a814d9ec07587cc9e41afa104aeecf175/pkg/geo/geogen/geogen.go#L63)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Was this not caught by a test because we later fix everything to be in CCW order?
</blockquote></details>

yep...


<!-- Sent from Reviewable.io -->

otan committed:
geogen: introduce package to generate random geo types

Generate random geo types in a new geogen package. This is immediately
utilised for all the randomized testing, including sqlsmith.

Release note: None

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51183#-:-MDK51Pp3uvZi6yFuLqy:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan, @rohany, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola
